Source/DafnyCore/Backends/Python/PythonCodeGenerator.cs

```diff
@@ -83,29 +99,107 @@ class PythonCodeGenerator : SinglePassCodeGenerator {
       return file;
     }
 
     protected override void DependOnModule(Program program, ModuleDefinition module, ModuleDefinition externModule,
       string libraryName) {
-      var moduleName = IdProtect(module.GetCompileName(Options));
-      Imports.Add(moduleName);
+      var dependencyPythonModuleName = "";
+
+      // If the module being depended on was compiled using module mode,
+      // this module needs to rely on it using the dependency's translation record,
+      // even if this module is not using module mode
+      var translatedRecord = program.Compilation.AlreadyTranslatedRecord;
+      translatedRecord.OptionsByModule.TryGetValue(module.FullDafnyName, out var moduleOptions);
+      object dependencyModuleName = null;
+      moduleOptions?.TryGetValue(PythonBackend.PythonModuleNameCliOption.Name, out dependencyModuleName);
+      if (dependencyModuleName is string && !string.IsNullOrEmpty((string)dependencyModuleName)) {
+        dependencyPythonModuleName = dependencyModuleName is string name ? (string)dependencyModuleName + "." : "";
+        if (String.IsNullOrEmpty(dependencyPythonModuleName)) {
+          Reporter.Warning(MessageSource.Compiler, ResolutionErrors.ErrorId.none, Token.Cli,
+            $"Python Module Name not found for the module {module.GetCompileName(Options)}");
+        }
+      }
+
+      var dependencyCompileName = IdProtect(module.GetCompileName(Options));
+      Imports.Add(dependencyPythonModuleName + dependencyCompileName, dependencyCompileName);
     }
 
-    private void EmitImports(string moduleName, ConcreteSyntaxTree wr) {
+    private void EmitImports(string moduleName, string pythonModuleName, ConcreteSyntaxTree wr) {
       wr.WriteLine("import sys");
       wr.WriteLine("from typing import Callable, Any, TypeVar, NamedTuple");
       wr.WriteLine("from math import floor");
       wr.WriteLine("from itertools import count");
       wr.WriteLine();
-      Imports.ForEach(module => wr.WriteLine($"import {module}"));
+      // Don't emit `import module_` for generated modules in the DafnyRuntimePython.
+      // The DafnyRuntimePython doesn't have a module.py file, so the import isn't valid.
+      if (!(DafnyRuntimeGeneratedModuleNames.Contains(moduleName))) {
+        wr.WriteLine($"import {(PythonModuleMode ? PythonModuleName + "." + DafnyDefaultModule : DafnyDefaultModule)} as {DafnyDefaultModule}");
+      }
+      foreach (var module in Imports) {
+        if (module.Value.Contains(".")) {
+          // This code branch only applies to extern modules. (ex. module {:extern "a.b.c"})
+          // If module.Value has `.`s in it, it is a nested extern module.
+          // (Non-extern nested modules with `.`s in their names
+          // will have their `.`s replaced with `_` before this code is executed,
+          // ex. module M.N --> "M_N", and will not hit this branch.)
+          // 
+          // Nested externs currently CANNOT be used with python-module-name:
```

Comment:
Thanks for the very detailed comment on this issue!

I think one of the factors making this worse is that `{:extern}` can be used either to say "this thing is implemented externally", or "this thing will be referenced by external code". In the latter case you're specifying the external name you want for the code you're generating, but in the former you're sometimes trying to declare the existing name for existing code. So whether `python-module-name` should apply to the extern name depends on the implicit use!

But to me that just means you're doing the right thing by not supporting either case for now - if your use cases can avoid it for now, that's the right call.